easynav_gps_vff/easynav_gps_vff_controller/src/easynav_gps_vff_controller/GpsVffController.cpp

```diff
@@ -0,0 +1,306 @@
+  marker.id = visualization_msgs::msg::Marker::ADD;
+
+  geometry_msgs::msg::Point start;
+  start.x = 0.0;
+  start.y = 0.0;
+  geometry_msgs::msg::Point end;
+  start.x = vector[0];
+  start.y = vector[1];
+  marker.points = {end, start};
+
+  marker.scale.x = 0.05;
+  marker.scale.y = 0.1;
+
+  switch (vff_color) {
+    case RED:
+      marker.id = 0;
+      marker.color.r = 1.0;
+      break;
+    case GREEN:
+      marker.id = 1;
+      marker.color.g = 1.0;
+      break;
+    case BLUE:
+      marker.id = 2;
+      marker.color.b = 1.0;
+      break;
+  }
+  marker.color.a = 1.0;
+
+  return marker;
+}
+
+double GpsVffController::normalize_angle(double angle)
+{
+  angle = fmod(angle + M_PI, 2.0 * M_PI); // range [0, 2π)
+  if (angle < 0) {
+    angle += 2.0 * M_PI; // ensure positive
+  }
+  return angle - M_PI;
+}
+
+VFFVectors GpsVffController::get_vff(
+  double angle_error,
+  const pcl::PointCloud<pcl::PointXYZ> & pointcloud_,
+  std::string frame_id)
+{
+  // Init vectors
+  VFFVectors vff_vector;
+  // Attractive vector points toward the goal
+  vff_vector.attractive = {std::cos(angle_error), std::sin(angle_error)};
+  vff_vector.repulsive = {0.0, 0.0};
+  vff_vector.result = {0.0, 0.0};
+
+  // Find the closest obstacle in the point cloud
+  pcl::PointXYZ closest_point;
+  double min_distance = std::numeric_limits<double>::max();
+
+  for (const auto & point : pointcloud_) {
+    const double distance = std::hypot(point.x, point.y);
+    if (distance < min_distance) {
+      min_distance = distance;
+      closest_point = point;
+    }
+  }
+
+  // Compute repulsive vector if an obstacle is within the threshold
+  if (min_distance < distance_obstacle_detection_) {
+    const double angle = std::atan2(closest_point.y, closest_point.x);
+    const double opposite_angle = angle + M_PI;
+    const double complementary_dist = distance_obstacle_detection_ - min_distance;
+
+    vff_vector.repulsive[0] = std::cos(opposite_angle) * complementary_dist;
+    vff_vector.repulsive[1] = std::sin(opposite_angle) * complementary_dist;
+  }
+
+  // Combine attractive and repulsive vectors
+  vff_vector.result[0] = vff_vector.attractive[0] + vff_vector.repulsive[0];
+  vff_vector.result[1] = vff_vector.attractive[1] + vff_vector.repulsive[1];
+
+  // Publish debug markers
+  visualization_msgs::msg::Marker marker;
+  marker.header.frame_id = cmd_vel_.header.frame_id;
+  marker.header.stamp = get_node()->now();
+  marker.type = visualization_msgs::msg::Marker::SPHERE;
+  marker.id = 456;
+  marker.pose.position.x = closest_point.x;
+  marker.pose.position.y = closest_point.y;
+  marker.pose.position.z = closest_point.z;
+  marker.pose.orientation.w = 1.0;
+  marker.scale.x = marker.scale.y = marker.scale.z = 0.1;
+  marker.color.a = 1.0;
+  marker.color.r = 1.0;
+  marker.color.g = 0.0;
+  marker.color.b = 0.0;
+
+  auto marker_array = get_debug_vff(vff_vector, frame_id);
+  marker_array.markers.push_back(marker);
+
```

Comment:
Maybe this should be inside the `if (marker_array_pub_->get_subscription_count() > 0) ` for avoiding extra computation if nobody is waiting for the markers.